src/wallet/witness_address.cpp

```diff
@@ -0,0 +1,550 @@
+            to_witness(witness, version_, witness_version_, witness_hash_);
+            return encode_base58(witness);
+        }
+
+        case witness_address::mainnet_bech32_p2wpkh:
+        case witness_address::mainnet_bech32_p2wsh:
+        case witness_address::testnet_bech32_p2wpkh:
+        case witness_address::testnet_bech32_p2wsh:
+            return bech32();
+
+        default:
+            return {};
+    }
+}
+
+// Accessors.
+// ----------------------------------------------------------------------------
+
+uint8_t witness_address::witness_version() const
+{
+    return witness_version_;
+}
+
+const hash_digest& witness_address::witness_hash() const
+{
+    return witness_hash_;
+}
+
+chain::script witness_address::output_script() const
+{
+    switch (version_)
+    {
+        case witness_address::mainnet_bech32_p2wpkh:
+        case witness_address::testnet_bech32_p2wpkh:
+        case witness_address::mainnet_base58_p2wpkh:
+        case witness_address::testnet_base58_p2wpkh:
+            return chain::script::to_pay_witness_key_hash_pattern(hash_);
+
+        case witness_address::mainnet_bech32_p2wsh:
+        case witness_address::testnet_bech32_p2wsh:
+        case witness_address::mainnet_base58_p2wsh:
+        case witness_address::testnet_base58_p2wsh:
+            return chain::script::to_pay_witness_script_hash_pattern(witness_hash_);
+
+        default:
+            return {};
+    }
+}
+
+// Methods.
+// ----------------------------------------------------------------------------
+
+/// static
+void witness_address::to_witness(witness_p2wpkh& out, uint8_t version,
+    uint8_t witness_version, short_hash hash)
+{
+    build_array(out,
+    {
+        to_array(version),
+        to_array(witness_version),
+        to_array(padding),
+        hash
+    });
+
+    insert_checksum(out);
+}
+
+/// static
+void witness_address::to_witness(witness_p2wsh& out, uint8_t version,
+    uint8_t witness_version, hash_digest hash)
+{
+    build_array(out,
+    {
+        to_array(version),
+        to_array(witness_version),
+        to_array(padding),
+        hash
+    });
+
+    insert_checksum(out);
+}
+
+// Operators.
+// ----------------------------------------------------------------------------
+
+witness_address& witness_address::operator=(const witness_address& other)
+{
+    valid_ = other.valid_;
+    version_ = other.version_;
+    hash_ = other.hash_;
+    prefix_ = other.prefix_;
+    witness_version_ = other.witness_version_;
+    witness_hash_ = other.witness_hash_;
+    return *this;
+}
+
+bool witness_address::operator<(const witness_address& other) const
+{
+    return encoded() < other.encoded();
+}
+
+bool witness_address::operator==(const witness_address& other) const
+{
+    return valid_ == other.valid_ && version_ == other.version_ &&
+        prefix_ == other.prefix_ &&
+        witness_version_ == other.witness_version_ && hash_ == other.hash_ &&
+        witness_hash_ == other.witness_hash_;
```

Comment:
This is the case where I prefer stacking (a long sequence of `&&` or `||` operations), an exception to the stacking style rule. 
```cpp
    return valid_ == other.valid_
        && version_ == other.version_
        && prefix_ == other.prefix_
        && witness_version_ == other.witness_version_
        && hash_ == other.hash_
        && witness_hash_ == other.witness_hash_;
```